Added AWS Lambda Node.js instructions (#5780)
elastic/apm-server: docs/guide/opentelemetry-elastic.asciidoc

layer]: The latest https://aws-otel.github.io/docs/getting-started/lambda/lambda-js[AWS
Lambda layer for OpenTelemetry]. For example, `arn:aws:lambda:eu-west-1:901920570463:layer:aws-otel-nodejs-ver-0-23-0:1`)
* https://docs.aws.amazon.com/lambda/latest/dg/API_TracingConfig.html[TracingConfig / Mode] set to `PassTrough`
* https://docs.aws.amazon.com/lambda/latest/dg/API_FunctionConfiguration.html[FunctionConfiguration / Timeout] set to more than 10 seconds to support the cold start of the Lambda JS Runtime
* Export the environment variables:
** `AWS_LAMBDA_EXEC_WRAPPER="/opt/otel-handler"` for wrapping handlers proxied through the API Gateway. See https://aws-otel.github.io/docs/getting-started/lambda/lambda-js#enable-auto-instrumentation-for-your-lambda-function[enable auto instrumentation for your lambda-function].
** `OTEL_PROPAGATORS="tracecontext"` to override the default setting that also enables X-Ray headers causing interferences between OpenTelemetry and X-Ray
** `OPENTELEMETRY_COLLECTOR_CONFIG_FILE="/var/task/opentelemetry-collector.yaml"` to specify the path to your OpenTelemetry Collector configuration

> ** `OTEL_EXPORTER_OTLP_ENDPOINT="http://localhost:55681/v1/traces"` this environment variable is required to be set until [ pull request](https://github.com/open-telemetry/opentelemetry-js/pull/2331) is merged and released.
> ** `OTEL_TRACES_SAMPLER="AlwaysOn"` define the required sampler strategy if it is not sent from the caller. Note that `Always_on` can potentially create a very large amount of data, so in production set the correct sampling configuration, as per [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampling).

Reviewer: Change from markdown to asciidoc links

```suggestion
** `OTEL_EXPORTER_OTLP_ENDPOINT="http://localhost:55681/v1/traces"` this environment variable is required to be set until https://github.com/open-telemetry/opentelemetry-js/pull/2331[PR #2331] is merged and released.
** `OTEL_TRACES_SAMPLER="AlwaysOn"` define the required sampler strategy if it is not sent from the caller. Note that `Always_on` can potentially create a very large amount of data, so in production set the correct sampling configuration, as per the https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampling[specification].
```

* The configuration of the OpenTelemetry Collector, with the definition of the Elastic Observability endpoint, can be added to the root directory of the Lambda binaries (e.g. defined in `src/main/resources/opentelemetry-collector.yaml`)
+
[source,yaml]
----
# Copy opentelemetry-collector.yaml in the root directory of the lambda function
# Set an environment variable 'OPENTELEMETRY_COLLECTOR_CONFIG_FILE' to '/var/task/opentelemetry-collector.yaml'
receivers:
  otlp:
    protocols:
      http:
      grpc:

exporters:
  logging:
    loglevel: debug
  otlp/elastic:
    # Elastic APM server https endpoint without the "https://" prefix
    endpoint: "${ELASTIC_OTLP_ENDPOINT}" <1>
    headers:
      # Elastic APM Server secret token
      Authorization: "Bearer ${ELASTIC_OTLP_TOKEN}" <1>

service:
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [logging, otlp/elastic]
    metrics:
      receivers: [otlp]
      exporters: [logging, otlp/elastic]
----
<1> Environment-specific configuration parameters can be conveniently passed in as environment variables: `ELASTIC_OTLP_ENDPOINT` and `ELASTIC_OTLP_TOKEN`

> * Configure the AWS Lambda Node.js function with:

Reviewer 2: ```suggestion
2. Configure the AWS Lambda Node.js function:

```

> ** https://docs.aws.amazon.com/lambda/latest/dg/API_Layer.html[Function
> layer]: The latest https://aws-otel.github.io/docs/getting-started/lambda/lambda-js[AWS
> Lambda layer for OpenTelemetry] (e.g. `arn:aws:lambda:eu-west-1:901920570463:layer:aws-otel-nodejs-ver-0-19-0:1`)

Reviewer 2: ```suggestion
* https://docs.aws.amazon.com/lambda/latest/dg/API_Layer.html[Function
layer]: The latest https://aws-otel.github.io/docs/getting-started/lambda/lambda-js[AWS
Lambda layer for OpenTelemetry]. For example, `arn:aws:lambda:eu-west-1:901920570463:layer:aws-otel-nodejs-ver-0-19-0:1`)
```

> ** https://docs.aws.amazon.com/lambda/latest/dg/API_TracingConfig.html[TracingConfig / Mode] set to `PassTrough`

Reviewer 2: ```suggestion
* https://docs.aws.amazon.com/lambda/latest/dg/API_TracingConfig.html[TracingConfig / Mode] set to `PassTrough`
```

> ** https://docs.aws.amazon.com/lambda/latest/dg/API_FunctionConfiguration.html[FunctionConfiguration / Timeout] set to more than 10 seconds to support the cold start of the Lambda JS Runtime

Reviewer 2: ```suggestion
* https://docs.aws.amazon.com/lambda/latest/dg/API_FunctionConfiguration.html[FunctionConfiguration / Timeout] set to more than 10 seconds to support the cold start of the Lambda JS Runtime
```

> ** Export the environment variables:

Reviewer 2: ```suggestion
* Export the environment variables:
```

> *** `AWS_LAMBDA_EXEC_WRAPPER="/opt/otel-handler"` for wrapping handlers proxied through the API Gateway (see https://aws-otel.github.io/docs/getting-started/lambda/lambda-js#enable-auto-instrumentation-for-your-lambda-function[here])

Reviewer 2: ```suggestion
** `AWS_LAMBDA_EXEC_WRAPPER="/opt/otel-handler"` for wrapping handlers proxied through the API Gateway. See https://aws-otel.github.io/docs/getting-started/lambda/lambda-js#enable-auto-instrumentation-for-your-lambda-function[enable auto instrumentation for your lambda-function].
```

Reviewer 3: @EamonnTP why do you want to remove the 3rd level indentation of the bullet list?

Reviewer 2: Hi @cyrille-leclerc This is to fall inline with my other suggestions above; dropping the bullets point down a level. There are still a few other comments that need addressed. Thanks.

> *** `OTEL_PROPAGATORS="tracecontext"` to override the default setting that also enables X-Ray headers causing interferences between OpenTelemetry and X-Ray

Reviewer 2: ```suggestion
** `OTEL_PROPAGATORS="tracecontext"` to override the default setting that also enables X-Ray headers causing interferences between OpenTelemetry and X-Ray
```

> *** `OPENTELEMETRY_COLLECTOR_CONFIG_FILE="/var/task/opentelemetry-collector.yaml"` to specify the path to your OpenTelemetry Collector configuration

Reviewer 2: ```suggestion
** `OPENTELEMETRY_COLLECTOR_CONFIG_FILE="/var/task/opentelemetry-collector.yaml"` to specify the path to your OpenTelemetry Collector configuration
```

[[open-telemetry-aws-lambda-elastic-nodejs-terraform]]
===== Instrumenting AWS Lambda Node.js functions with Terraform

> We recommend using an infrastructure as code solution like Terraform or Ansible to manage the configuration of your AWS Lambda functions.

Reviewer 2: ```suggestion
To manage the configuration of your AWS Lambda functions, we recommend using an infrastructure as code solution like Terraform or Ansible.
```

Here is an example of AWS Lambda Node.js function managed with Terraform and the https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_function[AWS Provider / Lambda Functions]:

> * Sample Terraform code: https://github.com/michaelhyatt/terraform-aws-nodejs-api-worker-otel/tree/v0.19

Reviewer 2: ```suggestion
* https://github.com/michaelhyatt/terraform-aws-nodejs-api-worker-otel/tree/v0.19[Sample Terraform code]
```

[[open-telemetry-aws-lambda-elastic-nodejs]]
===== Instrumenting AWS Lambda Node.js functions

NOTE: For a better startup time, we recommend using SDK-based instrumentation for manual instrumentation of the code rather than auto instrumentation.

To instrument AWS Lambda Node.js functions, see https://aws-otel.github.io/docs/getting-started/lambda/lambda-js[AWS Distro for OpenTelemetry Lambda Support For JS].

Noteworthy configuration elements:

IMPORTANT: Due to a https://github.com/aws-observability/aws-otel-lambda/issues/118[limitation] in AWS Lambda Otel Layer for Node.js v0.19, you must create a trace in the calling invocation of the Node.js lambda functions. This is to force the sampling strategy to `sampled:true` passing a `traceparent` header with `sampled=true`. For example, with `-01` at the end of the value.

To create a trace in the calling invocation of the Lambda function, use these or other methods that result in the `traceparent` header being added to the outgoing HTTP call:

> * If the caller is a web browser, instrument the caller using (https://www.elastic.co/guide/en/apm/get-started/current/rum.html)[Real User Monitoring (RUM)] by enabling https://www.elastic.co/guide/en/apm/agent/rum-js/5.x/distributed-tracing-guide.html[distributed tracing] on the RUM agent.

Reviewer 2: ```suggestion
* If the caller is a web browser, instrument the caller using <<rum,Real User Monitoring (RUM)>> by enabling {apm-rum-ref}/distributed-tracing-guide.html[distributed tracing] on the RUM agent.
```

* Sample Terraform code: https://github.com/cyrille-leclerc/my-serverless-shopping-cart/tree/main/checkout-function/deploy
* Note that the Terraform code to manage the HTTP API Gateway (https://github.com/cyrille-leclerc/my-serverless-shopping-cart/tree/main/utils/terraform/api-gateway-proxy[here]) is copied from the official OpenTelemetry Lambda sample https://github.com/open-telemetry/opentelemetry-lambda/tree/e72467a085a2a6e57af133032f85ac5b8bbbb8d1/utils[here]

[[open-telemetry-aws-lambda-elastic-nodejs]]
===== Instrumenting AWS Lambda Node.js functions

NOTE: For a better startup time, we recommend using SDK-based instrumentation for manual instrumentation of the code rather than auto instrumentation.

To instrument AWS Lambda Node.js functions, see https://aws-otel.github.io/docs/getting-started/lambda/lambda-js[AWS Distro for OpenTelemetry Lambda Support For JS].

> 
> Important configuration settings:

Reviewer: ```suggestion
```

IMPORTANT: Due to a https://github.com/aws-observability/aws-otel-lambda/issues/118[limitation] in AWS Lambda Otel Layer for Node.js v0.19, you must create a trace in the calling invocation of the Node.js lambda functions. This is to force the sampling strategy to `sampled:true` passing a `traceparent` header with `sampled=true`. For example, with `-01` at the end of the value.

To create a trace in the calling invocation of the Lambda function, use these or other methods that result in the `traceparent` header being added to the outgoing HTTP call:

* If the caller is a web browser, instrument the caller using <<rum,Real User Monitoring (RUM)>> by enabling {apm-rum-ref}/distributed-tracing-guide.html[distributed tracing] on the RUM agent.
* If the caller is an application, instrument the application with an OpenTelemetry or Elastic APM agent ensuring distributed tracing is enabled.

> 1. The configuration of the OpenTelemetry Collector, with the definition of the Elastic Observability endpoint, can be added to the root directory of the Lambda binaries: `src/main/resources/opentelemetry-collector.yaml`.
> +

Reviewer: ```suggestion
The configuration of the OpenTelemetry Collector, with the definition of the Elastic Observability endpoint, can be added to the root directory of the Lambda binaries: `src/main/resources/opentelemetry-collector.yaml`.

```

[source,yaml]
----
# Copy opentelemetry-collector.yaml in the root directory of the lambda function
# Set an environment variable 'OPENTELEMETRY_COLLECTOR_CONFIG_FILE' to '/var/task/opentelemetry-collector.yaml'
receivers:
  otlp:
    protocols:
      http:
      grpc:

exporters:
  logging:
    loglevel: debug
  otlp/elastic:
    # Elastic APM server https endpoint without the "https://" prefix
    endpoint: "${ELASTIC_OTLP_ENDPOINT}" <1>
    headers:
      # Elastic APM Server secret token
      Authorization: "Bearer ${ELASTIC_OTLP_TOKEN}" <1>

service:
  pipelines:
    traces:
      receivers: [otlp]
      exporters: [logging, otlp/elastic]
    metrics:
      receivers: [otlp]
      exporters: [logging, otlp/elastic]
----
<1> Environment-specific configuration parameters can be conveniently passed in as environment variables: `ELASTIC_OTLP_ENDPOINT` and `ELASTIC_OTLP_TOKEN`

> 2. Configure the AWS Lambda Node.js function:

Reviewer: ```suggestion
Configure the AWS Lambda Node.js function:
```
